### PR TITLE
LWDEV-9295 Ops Genie adapter job inital

### PR DESCRIPTION
**Pull request by some-dev**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,19 @@
 # Lykke.Job.OpsGenie
-OpsGenie external Service Integration
+OpsGenie (https://www.opsgenie.com/)  service external adapter, created for usage in Lykke infrastructure
+
+## Ops Genie Documentation
+https://docs.opsgenie.com/docs/alerts-page
+
+## What this service is inteded for
+Service provide  api to interact with OpsGenie api in queue based manner
```
Comment (KonstantinRyazantsev) [inline README.md]: Better to say "Service provide the messaging based API atop of OpsGenie API"

File: README.md [hunk 2]
```diff
@@ -1,2 +1,19 @@
...
+## How to use the service
+
+Include  https://www.nuget.org/packages/Lykke.Job.OpsGenie.Client/ in your project nuget package.
+### Usage examples
+- https://github.com/LykkeCity/Lykke.Job.OpsGenie/blob/dev/src/Lykke.Service.OpsGenieClientExample/Modules/ServiceModule.cs#L20 - client registration
+- https://github.com/LykkeCity/Lykke.Job.OpsGenie/blob/dev/src/Lykke.Service.OpsGenieClientExample/Controllers/AlertController.cs#L31 client usage
+
+## Lykke.Job.OpsGenie Configuration
+
+TBD
```
Comment (KonstantinRyazantsev) [inline README.md]: Can we add configuration docs now?

File: client/Lykke.Job.OpsGenie.Client/OpsGenieJobClient.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+using Lykke.Job.OpsGenie.Client.Mappers;
+using Lykke.Job.OpsGenie.Contract;
+
+namespace Lykke.Job.OpsGenie.Client
+{
+    /// <summary>
+    /// job client
+    /// </summary>
+    [PublicAPI]
+    public class OpsGenieJobClient:IOpsGenieJobClient
```
Comment (KonstantinRyazantsev) [inline client/Lykke.Job.OpsGenie.Client/OpsGenieJobClient.cs]: ":" should be sorrounded by spaces :)

File: src/Lykke.Job.OpsGenie.AzureRepositories/Domain/DomainEntity.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace Lykke.Job.OpsGenie.AzureRepositories.Domain
+{
+    internal class DomainEntity:TableEntity
+    {
+        public string Domain { get; set; }
+
+        public static string GenerateRowKey()
+        {
+            return "dmn";
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Job.OpsGenie.AzureRepositories/Domain/DomainEntity.cs]: Avoid usage of shortenings. If RK doesn't matter, better to use some meaningless symbol "-" for example.

File: src/Lykke.Job.OpsGenie.AzureRepositories/Lykke.Job.OpsGenie.AzureRepositories.csproj [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    <PackageReference Include="Lykke.AzureStorage" Version="8.11.0" />
+    <PackageReference Include="protobuf-net" Version="2.4.0" />
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\..\client\Lykke.Job.OpsGenie.Contract\Lykke.Job.OpsGenie.Contract.csproj" />
+    <ProjectReference Include="..\Lykke.Job.OpsGenie.Core\Lykke.Job.OpsGenie.Core.csproj" />
+  </ItemGroup>
+  <ItemGroup>
+    <Reference Include="protobuf-net">
+      <HintPath>C:\Users\NETSK\.nuget\packages\protobuf-net\2.4.0\lib\netcoreapp2.1\protobuf-net.dll</HintPath>
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Job.OpsGenie.AzureRepositories/Lykke.Job.OpsGenie.AzureRepositories.csproj]: Local machine specific hint path, better to remove it

File: src/Lykke.Job.OpsGenie.Core/Domain/Alert.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+        public string Entity { get; set; }
+
+        public string Alias { get; set; }
+
+        public string User { get; set; }
+
+        public string Source { get; set; }
+
+        public IDictionary<string, object> Details { get; set; }
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Job.OpsGenie.Core/Domain/Alert.cs]: IReadOnlyDictionary?

File: src/Lykke.Job.OpsGenie.Services/OpsGenieApi/Contracts/CreateAlertResponceContract.cs [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+{
+    internal class CreateAlertResponceContract
+    {
+        [JsonProperty("requestId")]
+        public string RequestId { get; set; }
+
+        [JsonProperty("took")]
+        public double Took { get; set; }
+
+        [JsonProperty("Result")]
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Job.OpsGenie.Services/OpsGenieApi/Contracts/CreateAlertResponceContract.cs]: Rest of field names are started from lower case. And do we really need all these JsonProperty? looks like they just match property names

File: src/Lykke.Job.OpsGenie.Services/QueueReader/DomainQueueReaderFactory.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            var apiAdapter = _apiAdapterStorage.GetOrDefault(queueMessage.Domain);
+
+            try
+            {
+                await apiAdapter.RaiseAlertAsync(queueMessage.MapFromQueueMessage());
+            }
+            catch (Exception e)
+            {
+                _logFactory.CreateLog(apiAdapter).Error(e, context: queueMessage);
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Job.OpsGenie.Services/QueueReader/DomainQueueReaderFactory.cs]: 1. Generally pattern is - create log in the ctor and use it the methods
2. And again, generally exceptions should be handled and logged at the top most level (by infrastructure). Will be good to add try/catch/log to the AzureQueueSubscriber and remove it from here

**Review by KonstantinRyazantsev — COMMENTED** [review]
Add description of the OpsGenieClient settings section to the task

File: src/Lykke.Job.OpsGenie/Settings/JobSettings/OpsGenieSettings.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Collections.Generic;
+using Lykke.SettingsReader.Attributes;
+
+namespace Lykke.Job.OpsGenie.Settings.JobSettings
+{
+    public class OpsGenieSettings
+    {
+        public string ApiUrl { get; set; }
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Job.OpsGenie/Settings/JobSettings/OpsGenieSettings.cs:8]: Can we add HttpCheck?

File: src/Lykke.Job.OpsGenie/Settings/SlackNotifications/AzureQueuePublicationSettings.cs [hunk 10]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Lykke.Job.OpsGenie.Settings.SlackNotifications
+{
+    public class AzureQueuePublicationSettings
+    {
+        public string ConnectionString { get; set; }
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Job.OpsGenie/Settings/SlackNotifications/AzureQueuePublicationSettings.cs:8]: AzureCheck?

File: src/Lykke.Job.OpsGenie/Startup.cs [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+
+        public void Configure(IApplicationBuilder app, IHostingEnvironment env, IApplicationLifetime appLifetime)
+        {
+            try
+            {
+                if (env.IsDevelopment())
+                    app.UseDeveloperExceptionPage();
+
+                app.UseLykkeForwardedHeaders();
+                app.UseLykkeMiddleware(ex => new ErrorResponse {ErrorMessage = "Technical problem"});
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Job.OpsGenie/Startup.cs]: Build the message from  the ex.ToAsyncString()

File: client/Lykke.Job.OpsGenie.Client/Alert.cs [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+        /// <summary>
+        /// Tags of the alert
+        /// </summary>
+        public ISet<string> Tags { get; }
+
+        /// <summary>
+        /// Map of key-value pairs to use as custom properties of the alert.
+        /// </summary>
+        public IDictionary<string, object> Details { get; }
```
Comment (KonstantinRyazantsev) [inline client/Lykke.Job.OpsGenie.Client/Alert.cs]: IReadOnlyDictionary?

File: src/Lykke.Job.OpsGenie/Settings/JobSettings/OpsGenieSettings.cs (hunk 9 above)
Comment (some-dev) [reply src/Lykke.Job.OpsGenie/Settings/JobSettings/OpsGenieSettings.cs:8]: This is ops genie api https://api.opsgenie.com/ - i cant find url, where 200 status code returned on GET request

File: src/Lykke.Job.OpsGenie/Startup.cs (hunk 11 above)
Comment (some-dev) [reply src/Lykke.Job.OpsGenie/Startup.cs]: Do we need to return full stacktrace here? Now its  omitted

Comment (KonstantinRyazantsev) [reply src/Lykke.Job.OpsGenie/Startup.cs]: Yep, this will help to investigate incidents

**Review by KonstantinRyazantsev — DISMISSED** [review]

**Review by KonstantinRyazantsev — APPROVED** [review]

**Pull request merged**